Sources/HummingbirdCore/Server/HTTP/HTTPChannelHandler.swift

```diff
@@ -43,57 +43,61 @@ extension HTTPChannelHandler {
-                                    await bodyStream.send(buffer)
-                                }
-                                bodyStream.finish()
-                            } catch {
-                                // pass failed to read full http body to request
-                                bodyStream.fail(error)
-                            }
-                            try await group.next()
-                            // set to idle unless it is cancelled then exit
-                            guard processingRequest.exchange(.idle, ordering: .relaxed) == .processing else { break }
+                try await asyncChannel.executeThenClose { inbound, outbound in
+                    let responseWriter = HBHTTPServerBodyWriter(outbound: outbound)
+                    var iterator = inbound.makeAsyncIterator()
+
+                    // read first part, verify it is a head
+                    guard let part = try await iterator.next() else { return }
+                    guard case .head(var head) = part else {
+                        throw HTTPChannelError.unexpectedHTTPPart(part)
+                    }
+
+                    while true {
+                        // set to processing unless it is cancelled then exit
+                        guard processingRequest.exchange(.processing, ordering: .relaxed) == .idle else { break }
+
+                        let bodyStream = HBStreamedRequestBody(iterator: iterator)
+                        let request = HBRequest(head: head, body: .stream(bodyStream))
+                        let response: HBResponse
+                        do {
+                            response = try await self.responder(request, asyncChannel.channel)
+                        } catch {
+                            response = self.getErrorResponse(from: error, allocator: asyncChannel.channel.allocator)
+                        }
+                        do {
+                            try await outbound.write(.head(response.head))
+                            let tailHeaders = try await response.body.write(responseWriter)
+                            try await outbound.write(.end(tailHeaders))
+                        } catch {
+                            throw error
+                        }
+                        if request.headers[.connection] == "close" {
+                            throw HTTPChannelError.closeConnection
                         }
+                        // set to idle unless it is cancelled then exit
+                        guard processingRequest.exchange(.idle, ordering: .relaxed) == .processing else { break }
+
+                        // Flush current request
+                        // read until we don't have a body part
+                        var part: HTTPRequestPart?
+                        while true {
+                            part = try await iterator.next()
+                            guard case .body = part else { break }
```

Comment:
We cannot iterate & discard here. If the user has completely drained their HTTPRequest, and reached `.end`, this next part will likely contain `.head(let newHead)`. We'll be discarding the new request, while attempting to get rid of unread body parts. We need to ensure that we've drained the entire body, but if we encounter a `.head`, we need to keep it and start handling the next request.